src/mrpro/algorithms/reconstruction/_direct_reconstruction.py

```diff
@@ -0,0 +1,122 @@
+        dcf
+            Density compensation. If None, no dcf will be performed.
+            Also set to None, if the FourierOperator is already density compensated.
+        """
+        super().__init__()
+        self.FourierOperator = fourier_operator
+        # TODO: Make this buffers once DataBufferMixin is merged
+        self.dcf = dcf
+        self.csm = csm
+        self.noise = noise
+
+    @classmethod
+    def from_kdata(cls, kdata: KData, noise: KNoise | None = None, coil_combine: bool = True) -> Self:
+        """Create a DirectReconstruction from kdata with default settings.
+
+        Parameters
+        ----------
+        kdata
+            KData to use for trajektory and header information
+        noise
+            KNoise used for whitening
+        coil_combine
+            if True (default), uses kdata to estimate sensitivity maps
+            and perform adaptive coil combine reconstruction
+        """
+        if noise is not None:
+            kdata = prewhiten_kspace(kdata, noise)
+        dcf = DcfData.from_traj_voronoi(kdata.traj)
+        fourier_op = FourierOp.from_kdata(kdata)
+        self = cls(fourier_op, None, noise, dcf)
+        if coil_combine:
+            self.recalculate_csm_walsh(kdata)
+        return self
+
+    def recalculate_fourierop(self, kdata: KData):
+        """Update the Fourier Operator, e.g. for a new trajectory.
+
+        Parameters
+        ----------
+        kdata
+            KData to determine trajectory and recon/encoding matrix from.
+        """
+        self.FourierOperator = FourierOp.from_kdata(kdata)
+        self.dcf = DcfData.from_traj_voronoi(kdata.traj)
+        return self
+
+    def recalculate_csm_walsh(self, kdata: KData, noise: KNoise | None = None):
+        """Update the CSM from KData using Walsh.
+
+        Parameters
+        ----------
+        kdata
+            KData used for adjoint reconstruction, which is then used for
+            Walsh CSM estimation.
+        noise
+            Noise measurement for prewhitening (optional)
+        """
+        adjoint = type(self)(self.FourierOperator, dcf=self.dcf, noise=noise)
+        image = adjoint(kdata)
+        self.csm = CsmData.from_idata_walsh(image)
+        return self
+
+    def forward(self, kdata: KData) -> IData:
+        """Apply the reconstruction.
+
+        Parameters
+        ----------
+        kdata
+            k-space data to reconstruct.
+
+        Returns
+        -------
+        image
+            the reconstruced image.
+        """
+        device = kdata.data.device
+        if self.noise is not None:
+            kdata = prewhiten_kspace(kdata, self.noise.to(device))
+        operator = self.FourierOperator
+        if self.csm is not None:
+            operator = operator @ self.csm.as_operator()
+        if self.dcf is not None:
+            operator = self.dcf.as_operator() @ operator
+        operator = operator.to(device)
+        (image_data,) = operator.H(kdata.data)
```

Comment:
should be img

Reply:
mhm.
```
        (img_data,) = operator.H(kdata.data)
        img = IData.from_tensor_and_kheader(img_data, kdata.header)
```
```
        (img,) = operator.H(kdata.data)
        image= IData.from_tensor_and_kheader(img_data, kdata.header)
```
```
        (img,) = operator.H(kdata.data)
        idata= IData.from_tensor_and_kheader(img, kdata.header)
```

the last one looks really strange to me, because the data of the image is named image, and the container is named data....

Reply:
an image-data-container contains an image and header information. to me this seems fine.

Reply:
but to call the variable idata does not really show what it is, that it is an image.. (except if you know about the idata class)

so img for the image would be better...
but img_data to me should be the actual image data, not the container...